(Docs) render strategies examples (#1120)
rx-angular/rx-angular: libs/cdk/render-strategies/docs/concurrent-strategies.md

# Concurrent Strategies

Based on the [RAIL model](https://web.dev/rail/), if your app provides user feedback within more than 16ms (less than 60 frames per second), it feels laggy to the user and leads to bad UX.
From the UX perspective that means users should not experience blocking periods more than 16 ms.

## Concepts
There are 5 core concepts of the concurrent strategies:
- Frame budget / Frame drop
- Scheduling
- Priority
- Chunking
- Concurrent Scheduling
### Frame budget / Frame Drop

The Browser has only one UI thread (main thread), meaning things happen one after another.

> Users constantly interract with out site and this means if our main thread is busy, they can't interact whit the page. The events like scroll or click will get delayed until the main thread is unblocked from work again and can process those interactions.

Author: ```suggestion
Users constantly interract with our site and this means if our main thread is busy, they can't interact whit the page. The events like scroll or click will get delayed until the main thread is unblocked from work again and can process those interactions.
```

<!-- TODO: Add proper image -->
![rx-angular-cdk-render-strategies__idle_example](https://user-images.githubusercontent.com/10064416/115316774-49225180-a17a-11eb-9045-3cdd38217b4d.PNG)

```typescript
@Component({
  selector: 'items-list',
  template: `
    <div id="items-list">
      <div *rxFor="let item of items$; strategy: 'normal'>
        {{item.name}}
      </div>
    </div>

    <button id="addItem" (click)="openCreateItemPopup()">Create new item</button>

    <div id="background-indicator>Background sync</div>
  `,
})
export class ItemsListComponent {
  items$ = this.state.items$;

  constructor(
    private state: StateService,
    private strategyProvider: RxStrategyProvider,
    private webSocket: WebSocketService
  ) {
    this.items$.pipe(
      this.strategyProvider.scheduleWith(items => this.webSocket.syncItems(items), {strategy: 'idle'})
    ).subscribe();
  }

  openCreateItemPopup() {
    this.strategyProvider.schedule(() => {
      // logic to lazy load popup component
    }, {strategy: 'low'}).subscribe();
  }

}
```

> **⚠ Notice:**  

> > This priority fits well for things that should happen but has lower priority. For any non-urgent background process `idle` is the best fit.

Author: ```suggestion
> This priority fits well for low priority background processes that are not affecting user interactions.
```

-
-->

Urgent work that should happen in the background and is not initiated but visible by the user. This occurs right after current task and has the lowest priority. 

| Render Method     | Scheduling    | Render Deadline |
| ----------------- | ------------- | --------------- |
| 🠗 `detectChanges` | `postMessage` | ❌              |

**Usecase:**

This strategy is especially useful for logic meant to run in the background. Good example of such interaction is background sync.

> Good use case for this strategy will be lazy load of component. For example popup.

Author: ```suggestion
```

<!-- TODO: Add proper image -->
![rx-angular-cdk-render-strategies__idle_example](https://user-images.githubusercontent.com/10064416/115316774-49225180-a17a-11eb-9045-3cdd38217b4d.PNG)

```typescript
@Component({
  selector: 'items-list',
  template: `
    <div id="items-list">
      <div *rxFor="let item of items$; strategy: 'normal'>
        {{item.name}}
      </div>
    </div>

    <button id="addItem" (click)="openCreateItemPopup()">Create new item</button>

    <div id="background-indicator>Background sync</div>
  `,
})
export class ItemsListComponent {
  items$ = this.state.items$;

  constructor(
    private state: StateService,
    private strategyProvider: RxStrategyProvider,
    private webSocket: WebSocketService
  ) {
    this.items$.pipe(

>       this.strategyProvider.scheduleWith(items => this.webSocket.syncItems(items))

Author: ```suggestion
      this.strategyProvider.scheduleWith(items => this.webSocket.syncItems(items), {strategy: 'idle'})
```
